src/databricks/labs/ucx/assessment/jobs.py

```diff
@@ -104,3 +118,233 @@ def snapshot(self) -> Iterable[JobInfo]:
     def _try_fetch(self) -> Iterable[JobInfo]:
         for row in self._fetch(f"SELECT * FROM {self._schema}.{self._table}"):
             yield JobInfo(*row)
+
+
+def dt_to_ms(dt: datetime):
+    epoch = datetime.utcfromtimestamp(0)
+    delta = dt - epoch
+    return int(delta.total_seconds() * 1000)
+
+
+def generate_hourly_ranges(start_date: datetime, end_date: datetime) -> list[tuple[datetime, datetime]]:
+    # Ensure start_date is earlier than end_date
+    if start_date > end_date:
+        raise ValueError("Start date should be earlier than end date")
+    iteration_date = start_date
+    hourly_dates = []
+
+    while iteration_date <= end_date:
+        hourly_dates.append((iteration_date, iteration_date + timedelta(hours=1)))
+        iteration_date += timedelta(hours=1)
+
+    return hourly_dates
+
+
+def _get_job_cluster_from_task(
+    task: RunTask, job_run: BaseRun, all_clusters: dict[str, ClusterDetails]
+) -> compute.ClusterSpec | ClusterDetails | None:
+    """
+    Determine the cluster associated with the task
+    1) Look for new_cluster on the task
+    2) Look for existing cluster on the task
+    Note - this function has DIFFERENT logic from the _get_cluster_configs_from_all_jobs
```

Comment:
Refactor JobsMixin to use the same logic

Reply:
It's a **different** logic. What's the point of generalizing it, when it's **essentially and logically** not the same?

Reply:
re-used logic where possible